File: research/explorations/bench-realistic/matmul.go
package main

import (
	"fmt"
	"os"
	"strconv"
	"time"
)

func bench(n int64) int64 {
	var s int64
	for i := int64(0); i < n; i++ {
		for j := int64(0); j < 10; j++ {
			for k := int64(0); k < 10; k++ {
				s += (i + j) * (j + k)
			}
		}
	}
	return s
}

func main() {
	n := int64(1000)
	if len(os.Args) > 1 {
		if v, err := strconv.ParseInt(os.Args[1], 10, 64); err == nil {
			n = v
		}
	}
	for i := 0; i < 1000; i++ {
		bench(n)
	}
	iters := int64(10000)
	start := time.Now()
	var r int64
	for i := int64(0); i < iters; i++ {
		r = bench(n)
	}
	elapsed := time.Since(start).Nanoseconds()
	fmt.Printf("result:     %d\n", r)
	fmt.Printf("iterations: %d\n", iters)
	fmt.Printf("total:      %.2fms\n", float64(elapsed)/1e6)
	fmt.Printf("per call:   %dns\n", elapsed/iters)
}

File: research/explorations/bench-realistic/matmul.lua
local function bench(n)
    local s = 0
    for i = 0, n - 1 do
        for j = 0, 9 do
            for k = 0, 9 do
                s = s + (i + j) * (j + k)
            end
        end
    end
    return s
end

local N = tonumber(arg[1]) or 1000
for i = 1, 1000 do bench(N) end
local iters = 10000
local clock = os.clock
local start = clock()
local r
for i = 1, iters do r = bench(N) end
local elapsed = (clock() - start) * 1e9
io.write(string.format("result:     %d\n", r))
io.write(string.format("iterations: %d\n", iters))
io.write(string.format("total:      %.2fms\n", elapsed / 1e6))
io.write(string.format("per call:   %.0fns\n", elapsed / iters))
